aten/src/THC/THCAtomics.cuh

```diff
@@ -12,36 +12,46 @@ struct AtomicAddIntegerImpl;
 template<typename T>
 struct AtomicAddIntegerImpl<T, 1> {
   inline __device__ void operator()(T *address, T val) {
-    uint32_t * address_as_ui =
-        (uint32_t *) (address - ((size_t)address & 3));
+    size_t offset = (size_t)address & 3;
+    uint32_t * address_as_ui = (uint32_t *)((char *)address - offset);
     uint32_t old = *address_as_ui;
-    uint32_t shift = (((size_t)address & 3) * 8);
-    uint32_t sum;
+    uint32_t shift = offset * 8;
+    uint32_t old_bytes;
+    uint32_t newval;
     uint32_t assumed;
 
     do {
       assumed = old;
-      sum = val + T((old >> shift) & 0xff);
-      old = (old & ~(0x000000ff << shift)) | (sum << shift);
-      old = atomicCAS(address_as_ui, assumed, old);
+      old_bytes = (old >> shift) & 0xff;
+      T tmp = THCNumerics<T>::add(val, old_bytes);
+      // maintain same size in initial cast to avoid padding negative values with 1s
```

Comment:
As the code reader I'd like a little more information. Something like:
"If we did something like uint32_t newval = tmp (which is -1), then newval = 0xffffffff, but we really just wanted to store one byte (0x000000ff)."

Reply:
updated a bit, also made the cast in-line instead of extra tmp values.